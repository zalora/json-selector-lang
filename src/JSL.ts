// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import Lexer from './Lexer';
import Program from './AST/Program';
import Parser from './Parser';

// JSL parses raw JSL input and returns a program that can be evaluated.
class JSL {
  static async compile(input: string): Promise<Program> {
    const lexer = new Lexer(input);
    const parser = new Parser(lexer);

    return parser.parseProgram();
  }
}

export default JSL;
