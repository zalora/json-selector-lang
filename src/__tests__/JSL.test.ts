// Copyright (c) ZALORA
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import JSL from '../JSL';

describe('JSL', () => {
  it('tests with valid & complex jsl input', () => {
    const input = '.data[224].width';
    const data = [];
    data.length = 230;
    data[224] = { width: 20 };
    const json = { data };

    try {
      const program = JSL.compile(input);
      const width = JSL.evaluate(json, program);

      expect(program).toMatchSnapshot();
      expect(width).toMatchSnapshot();
    } catch (e: any) {
      console.log(e);
      fail();
    }
  });

  it('tests with valid jsl input', () => {
    const input = '.data[2].width';
    const json = { data: [{}, {}, { width: 200 }] };

    try {
      const program = JSL.compile(input);
      const width = JSL.evaluate(json, program);

      expect(program).toMatchSnapshot();
      expect(width).toMatchSnapshot();
    } catch (e: any) {
      console.log(e);
      fail();
    }
  });

  it('tests with invalid jsl input', () => {
    const input = '..[]';
    let returnedError: Array<string> = [];

    try {
      JSL.compile(input);
    } catch (e: any) {
      returnedError = e;
    }

    if (returnedError.length < 0) {
      fail();
    }

    expect(returnedError.length).toEqual(4);
    const testCases = [
      'expected next token to be Symbol(ident), got Symbol(.) instead',
      'expected next token to be Symbol(ident), got Symbol([) instead',
      'expected next token to be Symbol(int), got Symbol(]) instead',
      'prefix parse func for ] not found',
    ];
    testCases.forEach((testCase, idx) => {
      expect(returnedError[idx]).toEqual(testCase);
    });
  });
});
