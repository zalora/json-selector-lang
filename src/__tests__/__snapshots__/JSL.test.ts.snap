// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSL tests with empty jsl input 1`] = `
Program {
  "statements": Array [],
}
`;

exports[`JSL tests with valid & complex jsl input 1`] = `
Program {
  "statements": Array [
    ASTExpressionStatement {
      "expression": ASTIndexExpression {
        "index": ASTIntegerLiteral {
          "token": Object {
            "literal": "224",
            "type": Symbol(int),
          },
          "value": 224,
        },
        "left": ASTSelectExpression {
          "key": "data",
          "token": Object {
            "literal": ".",
            "type": Symbol(.),
          },
        },
        "token": Object {
          "literal": "[",
          "type": Symbol([),
        },
      },
      "token": Object {
        "literal": ".",
        "type": Symbol(.),
      },
    },
    ASTExpressionStatement {
      "expression": ASTSelectExpression {
        "key": "width",
        "token": Object {
          "literal": ".",
          "type": Symbol(.),
        },
      },
      "token": Object {
        "literal": ".",
        "type": Symbol(.),
      },
    },
  ],
}
`;

exports[`JSL tests with valid & complex jsl input 2`] = `20`;

exports[`JSL tests with valid jsl input 1`] = `
Program {
  "statements": Array [
    ASTExpressionStatement {
      "expression": ASTIndexExpression {
        "index": ASTIntegerLiteral {
          "token": Object {
            "literal": "2",
            "type": Symbol(int),
          },
          "value": 2,
        },
        "left": ASTSelectExpression {
          "key": "data",
          "token": Object {
            "literal": ".",
            "type": Symbol(.),
          },
        },
        "token": Object {
          "literal": "[",
          "type": Symbol([),
        },
      },
      "token": Object {
        "literal": ".",
        "type": Symbol(.),
      },
    },
    ASTExpressionStatement {
      "expression": ASTSelectExpression {
        "key": "width",
        "token": Object {
          "literal": ".",
          "type": Symbol(.),
        },
      },
      "token": Object {
        "literal": ".",
        "type": Symbol(.),
      },
    },
  ],
}
`;

exports[`JSL tests with valid jsl input 2`] = `200`;
